src/daft-scan/src/glob.rs

```diff
@@ -12,28 +15,68 @@ use crate::{
 };
 #[derive(Debug, PartialEq, Hash)]
 pub struct GlobScanOperator {
-    glob_path: String,
+    glob_paths: Vec<String>,
     file_format_config: Arc<FileFormatConfig>,
     schema: SchemaRef,
     storage_config: Arc<StorageConfig>,
 }
 
+/// Wrapper struct that implements a sync Iterator for a tokio::sync::mpsc::Receiver
+struct TokioMPSCReceiverIterator<T> {
+    recv: Receiver<T>,
+}
+impl<T> Iterator for TokioMPSCReceiverIterator<T> {
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.recv.blocking_recv()
+    }
+}
+
+#[derive(Snafu, Debug)]
+enum Error {
+    #[snafu(display("Glob path had no matches: \"{}\"", glob_path))]
+    GlobNoMatch { glob_path: String },
+}
+
+impl From<Error> for DaftError {
+    fn from(value: Error) -> Self {
+        match &value {
+            Error::GlobNoMatch { glob_path } => DaftError::FileNotFound {
+                path: glob_path.clone(),
+                source: Box::new(value),
+            },
+        }
+    }
+}
+
 fn run_glob(
     glob_path: &str,
     limit: Option<usize>,
     io_client: Arc<IOClient>,
     runtime: Arc<tokio::runtime::Runtime>,
-) -> DaftResult<Vec<String>> {
+) -> DaftResult<Box<dyn Iterator<Item = DaftResult<String>>>> {
     let (_, parsed_glob_path) = parse_url(glob_path)?;
     let _rt_guard = runtime.enter();
+
     runtime.block_on(async {
-        Ok(io_client
+        let (sender, recv) = tokio::sync::mpsc::channel(128);
+        let mut fm_stream = io_client
             .as_ref()
             .glob(&parsed_glob_path, None, None, limit, None)
-            .await?
-            .into_iter()
-            .map(|fm| fm.filepath)
-            .collect())
+            .await?;
+
+        // Spawn a task to asynchronously send data into the spawned channel with best-effort
+        tokio::spawn(async move {
```

Comment:
this looks like it will delete the glob stream even if the iterator goes out of scope. I was thinking something more along the lines of [this](https://github.com/Eventual-Inc/Daft/blob/main/src/daft-parquet/src/file.rs#L50). but you shouldn't need the Mutex or curr. it should just look something like
```
    fn next(&mut self) -> Option<Self::Item> {
       self.handle.block_on(async {
         self.stream.next()
       })
    }


```

Reply:
> this looks like it will delete the glob stream even if the iterator goes out of scope

What do you mean by this?

Reply:
whoops typo, meant to say deplete. eg lets say you only pluck a few items out of the iterator, then you drop it. I believe this code will still hold onto the channel and stream since the tokio task is in the global lifetime. We want the behavior of dropping the iterator will immediately drop the stream.

Reply:
I see, I was thinking it would look something like:

1. Pluck a few items from iterator, iterator is dropped
2. The channel `recv` end is closed, and the `send` end will throw an error
3. The tokio task will exit

I'm not sure what happens to the lifetime of the `Runtime` object in this scenario though 😬 

I can swtich over to holding a Runtime instead of that seems cleaner. Wonder if there may be significant overhead in calling `self.handle.block_on` in a tight loop though, I'll check.

Reply:
Changed to use a loop on a handle